### PR TITLE
fix: cancel fetchComponent in BeagleViewModel

### LINKED ISSUE #983: Loading on Beagle Activity does not close when triggering onBackPressed
## Description

Please provide a clear and concise description of what the bug is. Include screenshots if needed.
Please test using the latest Beagle release to make sure your issue has not already been fixed: https://github.com/ZupIT/beagle/releases

## Steps To Reproduce

In the Android sample application, in the navigation part, when calling loadView from a remote screen, the load appears. However, when clicking on the Android back button, we return to the previous screen, but the load still continues. Another detail would be that when the load still appears we can click on components of the screen even the load on top.

## Expected Results

I believe that when the load is on the screen, there are 2 possibilities for this problem:
- Do not allow the user to press the back button;
- When the user clicks on return, change the state of the screen that is loading so that the listener changes the state of the screen to loaded;

I think the second option is better.

### PR DESCRIPTION
**Pull request by FredericoNoce**
### Related Issues

Closes #983 

### Description and Example

When the user triggers onBackPressed on BeagleActivity, we now validate that a fetchComponent still exists. If it exists, we cancel this fetch and remove the load from the screen.

### Checklist

Please, check if these important points are met using `[x]`:

- [x] I read the [PR Guide] and followed the process outlined there for submitting this PR.
- [x] I avoided _breaking changes_ by not changing public APIs that people rely on. 
- [x] I am willing to follow-up on review comments in a timely manner.

[PR Guide]: https://github.com/ZupIT/beagle/blob/master/doc/contributing/pull_requests.md

### CODE REVIEW

File: android/beagle/src/main/java/br/com/zup/beagle/android/view/viewmodel/BeagleViewModel.kt [hunk 1]
```diff
@@ -28,26 +28,29 @@ import br.com.zup.beagle.android.utils.CoroutineDispatchers
 import br.com.zup.beagle.android.view.ScreenRequest
 import br.com.zup.beagle.core.IdentifierComponent
 import br.com.zup.beagle.core.ServerDrivenComponent
-import kotlinx.coroutines.CoroutineDispatcher
-import kotlinx.coroutines.CoroutineScope
-import kotlinx.coroutines.launch
-import kotlinx.coroutines.withContext
+import kotlinx.coroutines.*
```
Comment (botbeaglezup) [inline android/beagle/src/main/java/br/com/zup/beagle/android/view/viewmodel/BeagleViewModel.kt:31]: <table data-meta="generated_by_android">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/ZupIT/beagle/blob/5e29abe5476db5489d8b6396b7dd207771ce36f2/android/beagle/src/main/java/br/com/zup/beagle/android/view/viewmodel/BeagleViewModel.kt#L31"></span>kotlinx.coroutines.* is a wildcard import. Replace it with fully qualified imports.</td>
    </tr>
  </tbody>
</table>

File: android/beagle/src/main/java/br/com/zup/beagle/android/view/BeagleActivity.kt [hunk 2]
```diff
@@ -227,7 +232,11 @@ abstract class BeagleActivity : AppCompatActivity() {
...
 
     override fun onBackPressed() {
-        if (supportFragmentManager.backStackEntryCount == 1) {
+        if (viewModel.isFetchComponent()) {
+            if (supportFragmentManager.backStackEntryCount == 0) {
+                finish()
+            }
+        } else if (supportFragmentManager.backStackEntryCount == 1) {
             finish()
         } else {
```
Comment (diogooliveira-zup) [inline android/beagle/src/main/java/br/com/zup/beagle/android/view/BeagleActivity.kt:241]: Hi! What you think to change this if statement to when?

File: android/beagle/src/test/java/br/com/zup/beagle/android/view/viewmodel/BeagleViewModelTest.kt [hunk 3]
```diff
@@ -29,23 +30,16 @@ import br.com.zup.beagle.android.testutil.RandomData
...
-import io.mockk.Runs
-import io.mockk.coEvery
-import io.mockk.coVerifyOrder
-import io.mockk.coVerifySequence
-import io.mockk.every
+import io.mockk.*
 import io.mockk.impl.annotations.MockK
-import io.mockk.just
-import io.mockk.mockk
-import io.mockk.verify
```
Comment (diogooliveira-zup) [inline android/beagle/src/test/java/br/com/zup/beagle/android/view/viewmodel/BeagleViewModelTest.kt:41]: Maybe is better to avoid wild card import

**Review by diogooliveira-zup — DISMISSED** [review]

File: android/beagle/src/main/java/br/com/zup/beagle/android/view/BeagleActivity.kt (hunk 2 above)
Comment (FredericoNoce) [reply android/beagle/src/main/java/br/com/zup/beagle/android/view/BeagleActivity.kt:241]: I believe so, but is there a need?

File: android/beagle/src/test/java/br/com/zup/beagle/android/view/viewmodel/BeagleViewModelTest.kt (hunk 3 above)
Comment (FredericoNoce) [reply android/beagle/src/test/java/br/com/zup/beagle/android/view/viewmodel/BeagleViewModelTest.kt:41]: ok!

File: android/beagle/src/main/java/br/com/zup/beagle/android/view/ServerDrivenActivity.kt [hunk 4]
```diff
@@ -37,29 +37,41 @@ class ServerDrivenActivity : BeagleActivity() {
...
+                    Snackbar.make(server_driven_container, R.string.beagle_label_generic_error, Snackbar.LENGTH_LONG)
+                        .setAction(R.string.beagle_label_try_again) { state.retry() }
+                        .show()
+                } else {
+                    buttonRetry.setOnClickListener {
+                        state.retry()
+                    }
+                    server_driven_container.visibility = View.GONE
+                    include_layout_error.visibility = View.VISIBLE
+                }
```
Comment (uziasferreirazup) [inline android/beagle/src/main/java/br/com/zup/beagle/android/view/ServerDrivenActivity.kt:57]: What do you think about a return this code to a function?
```suggestion
                handleError() or showGenericError()
```

File: android/beagle/src/test/java/br/com/zup/beagle/android/view/viewmodel/BeagleViewModelTest.kt [hunk 5]
```diff
@@ -201,4 +193,66 @@ class BeagleViewModelTest {
...
+        val mockJob = Job()
+
+        beagleUIViewModel.fetchComponent(screenRequest).observeForever(observer)
+        beagleUIViewModel.fetchComponent?.job = mockJob
+
+        // When
+        val isFetch = beagleUIViewModel.isFetchComponent()
+
+        //THEN
+        Assert.assertEquals(mockJob.isCancelled, true)
```
Comment (uziasferreirazup) [inline android/beagle/src/test/java/br/com/zup/beagle/android/view/viewmodel/BeagleViewModelTest.kt:254]:
```suggestion
     assertTrue(mockJob.isCancelled)
```

File: android/beagle/src/test/java/br/com/zup/beagle/android/view/viewmodel/BeagleViewModelTest.kt [hunk 6]
```diff
@@ -201,4 +193,66 @@ class BeagleViewModelTest {
...
+        verify(exactly = once()) { observer.onChanged(ViewState.DoCancel) }
+        Assert.assertEquals(isFetch, true)
```
Comment (uziasferreirazup) [inline android/beagle/src/test/java/br/com/zup/beagle/android/view/viewmodel/BeagleViewModelTest.kt:256]:
```suggestion
       assertTrue(isFetch)
```

File: android/beagle/src/test/java/br/com/zup/beagle/android/view/viewmodel/BeagleViewModelTest.kt [hunk 7]
```diff
@@ -201,4 +193,66 @@ class BeagleViewModelTest {
...
+        val screenRequest = ScreenRequest("")
+
+        beagleUIViewModel.fetchComponent(screenRequest)
+        beagleUIViewModel.fetchComponent?.job = null
+
+        // When
+        val isFetch = beagleUIViewModel.isFetchComponent()
+
+        //THEN
+        Assert.assertEquals(isFetch, false)
```
Comment (uziasferreirazup) [inline android/beagle/src/test/java/br/com/zup/beagle/android/view/viewmodel/BeagleViewModelTest.kt:221]:
```suggestion
        assertFalse(isFetch)
```

File: android/beagle/src/test/java/br/com/zup/beagle/android/view/viewmodel/BeagleViewModelTest.kt (hunk 3 above)
Comment (uziasferreirazup) [reply android/beagle/src/test/java/br/com/zup/beagle/android/view/viewmodel/BeagleViewModelTest.kt:41]: @diogooliveira-zup the validation of lint not applied to tests, you can use a wild card import.

File: android/beagle/src/main/java/br/com/zup/beagle/android/view/viewmodel/BeagleViewModel.kt [hunk 8]
```diff
@@ -94,6 +106,29 @@ internal class BeagleViewModel(
             }
         }
 
+        fun checkFetchComponent(): Boolean {
+            job?.let {
+                return fetchComponentIsCompleted(it)
+            }
+
+            return false
```
Comment (uziasferreirazup) [inline android/beagle/src/main/java/br/com/zup/beagle/android/view/viewmodel/BeagleViewModel.kt:114]:
```suggestion
           return job?.let {
                 fetchComponentIsCompleted(it)
            } ?: false
```

File: android/beagle/src/main/java/br/com/zup/beagle/android/view/BeagleActivity.kt (hunk 2 above)
Comment (diogooliveira-zup) [reply android/beagle/src/main/java/br/com/zup/beagle/android/view/BeagleActivity.kt:241]: No, it`s just a suggestion to decrease the code chars count

File: android/beagle/src/test/java/br/com/zup/beagle/android/view/viewmodel/BeagleViewModelTest.kt (hunk 3 above)
Comment (diogooliveira-zup) [reply android/beagle/src/test/java/br/com/zup/beagle/android/view/viewmodel/BeagleViewModelTest.kt:41]: okay, so this is resolved

File: android/beagle/src/main/java/br/com/zup/beagle/android/view/viewmodel/BeagleViewModel.kt [hunk 9]
```diff
@@ -94,6 +106,29 @@ internal class BeagleViewModel(
...
+        }
+
+        private fun fetchComponentIsCompleted(job: Job): Boolean {
+            return if (!job.isCompleted) {
+                cancelFetchComponent(job)
+                true
+            } else
+                false
```
Comment (uziasferreirazup) [inline android/beagle/src/main/java/br/com/zup/beagle/android/view/viewmodel/BeagleViewModel.kt:122]:
```suggestion
            val isCompleted = !job.isCompleted
            if (isCompleted) {
                cancelFetchComponent(job)
            }
            return isCompleted
```

**Review by uziasferreirazup — APPROVED** [review]

**Review by paulomeurerzup — APPROVED** [review]

**Review by uziasferreirazup — APPROVED** [review]

**Pull request merged**